cirq-google/cirq_google/engine/stream_manager.py

```diff
@@ -59,5 +80,222 @@ def publish_exception(self, exception: GoogleAPICallError) -> None:
+            StreamError if there is a non-retryable error while executing the job.
+            concurrent.futures.CancelledError if the stream is stopped while a job is in flight.
+            google.api_core.exceptions.GoogleAPICallError if the stream breaks with a non-retryable
+                error.
+        """
+        if self._manage_stream_loop_future is None:
+            self._manage_stream_loop_future = self._executor.submit(self._manage_stream)
+        return self._executor.submit(self._manage_execution, project_name, program, job)
+
+    def stop(self) -> None:
+        """Closes the open stream and resets all management resources."""
+        if self._manage_stream_loop_future is not None:
+            self._manage_stream_loop_future.cancel()
+        self._reset()
+
+    def _reset(self):
+        """Resets the manager state."""
+        self._request_queue: asyncio.Queue = asyncio.Queue()
+        self._manage_stream_loop_future: Optional[duet.AwaitableFuture] = None
+        # TODO(#5996) consider making the scope of response futures local to the relevant tasks
+        # rather than all of StreamManager.
+        self._response_demux = ResponseDemux()
+        self._next_available_message_id = 0
+
+    @property
+    def _executor(self) -> AsyncioExecutor:
+        # We must re-use a single Executor due to multi-threading issues in gRPC
+        # clients: https://github.com/grpc/grpc/issues/25364.
+        return AsyncioExecutor.instance()
+
+    # TODO(#5996) Add stop signal to the request iterator.
+    async def _request_iterator(self) -> AsyncIterator[quantum.QuantumRunStreamRequest]:
+        """The request iterator for Quantum Engine client RPC quantum_run_stream().
+
+        Every call to this method generates a new iterator.
+        """
+        while True:
+            yield await self._request_queue.get()
+
+    async def _manage_stream(self):
+        """The stream coroutine, an asyncio coroutine to manage QuantumRunStream.
+
+        This coroutine reads responses from the stream and forwards them to the ResponseDemux, where
+        the corresponding execution coroutine `_manage_request()` is notified.
+
+        When the stream breaks, the stream is reopened, and all execution coroutines are notified.
+
+        There is at most a single instance of this coroutine running.
+        """
+        while True:
+            try:
+                # The default gRPC client timeout is used.
+                response_iterable = await self._grpc_client.quantum_run_stream(
+                    self._request_iterator()
+                )
+                async for response in response_iterable:
+                    self._response_demux.publish(response)
+            except BaseException as e:
+                # TODO(#5996) Close the request iterator to close the existing stream.
+                # Note: the message ID counter is not reset upon a new stream.
```

Comment:
Thanks for the catch. We actually probably shouldn't reset the ID to avoid potential for any collisions. For example, the response to a request from a previous stream may be sent over the new stream.